<a href="https://www.kaggle.com/elakapoor/bar-charts-with-plotly?scriptVersionId=89848115" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

import plotly.express as px

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/geeksforgeeks-articles/articles.csv


In [2]:
df = pd.read_csv("/kaggle/input/geeksforgeeks-articles/articles.csv")
df.head()

title         author_id  \
0        5 Best Practices For Writing SQL Joins       priyankab14   
1                  Foundation CSS Dropdown Menu  ishankhandelwals   
2  Top 20 Excel Shortcuts That You Need To Know       priyankab14   
3                     Servlet – Fetching Result   nishatiwari1719   
4                              Suffix Sum Array          rohit768   

   last_updated                                               link category  
0  21 Feb, 2022  https://www.geeksforgeeks.org/5-best-practices...     easy  
1  20 Feb, 2022  https://www.geeksforgeeks.org/foundation-css-d...     easy  
2  17 Feb, 2022  https://www.geeksforgeeks.org/top-20-excel-sho...     easy  
3  17 Feb, 2022  https://www.geeksforgeeks.org/servlet-fetching...     easy  
4  21 Feb, 2022    https://www.geeksforgeeks.org/suffix-sum-array/     easy

In [3]:
df["year"] = df["last_updated"].str.split(",", expand = True).get(1)
df["date_month"] = df["last_updated"].str.split(",", expand = True).get(0)
df["month"] = df["date_month"].str.split(" ", expand = True).get(1)
df["date"] = df["date_month"].str.split(" ", expand = True).get(0)

In [4]:
df.drop(["link"], axis = 1, inplace = True)

In [5]:
df.head()

title         author_id  \
0        5 Best Practices For Writing SQL Joins       priyankab14   
1                  Foundation CSS Dropdown Menu  ishankhandelwals   
2  Top 20 Excel Shortcuts That You Need To Know       priyankab14   
3                     Servlet – Fetching Result   nishatiwari1719   
4                              Suffix Sum Array          rohit768   

   last_updated category   year date_month month date  
0  21 Feb, 2022     easy   2022     21 Feb   Feb   21  
1  20 Feb, 2022     easy   2022     20 Feb   Feb   20  
2  17 Feb, 2022     easy   2022     17 Feb   Feb   17  
3  17 Feb, 2022     easy   2022     17 Feb   Feb   17  
4  21 Feb, 2022     easy   2022     21 Feb   Feb   21

In [6]:
df.shape

(34574, 8)

In [7]:
null_index = df.isnull().sum().index
null_values = df.isnull().sum().values
null_values

array([  0,  19,  18,   0, 114,  18, 114,  18])

# Basic Charts 

In [8]:
# bar charts
fig = px.bar(x = null_index, y = null_values) # providing x and y values

fig.update_layout(
        xaxis_title = "Column Name", # x-axis label
        yaxis_title = "Missing Values", # y-axis label
        title = "Total missing values per column") # bar plot title
 
fig.show()

Conclusions:
1. 19 author_id are missing
2. last_updated have 18 missing values
3. date extracted from the last_updated also have 18 missing values
4. Column year and month have 114 missing values. These columns are extracted from the last_updated column.

In [9]:
df[df['year'].isna()]

title      author_id  \
20     Must Do Coding Questions for Product Based Com...  GeeksforGeeks   
152    Get Hired With GeeksforGeeks and Win Exciting ...  GeeksforGeeks   
617    Recently Asked Interview Questions in Product ...  GeeksforGeeks   
654    100 Days of Code – A Complete Guide For Beginn...    anuupadhyay   
711    Be a Part of GeeksforGeeks YouTube World – Mor...  GeeksforGeeks   
...                                                  ...            ...   
31290                  Brent’s Cycle Detection Algorithm     Surya Priy   
31348                                       Sparse Table    pawan_asipu   
31567   Number Theory (Interesting Facts and Algorithms)  GeeksforGeeks   
31570                          Magic Square | Even Order  GeeksforGeeks   
31596                               Bidirectional Search  GeeksforGeeks   

      last_updated category  year date_month month    date  
20          Medium     easy  None     Medium  None  Medium  
152           Easy     easy  None       Easy  None    Easy  
617         Medium     easy  None     Medium  None  Medium  
654         Medium     easy  None     Medium  None  Medium  
711         Medium     easy  None     Medium  None  Medium  
...            ...      ...   ...        ...   ...     ...  
31290         Hard     hard  None       Hard  None    Hard  
31348         Hard     hard  None       Hard  None    Hard  
31567         Hard     hard  None       Hard  None    Hard  
31570         Hard     hard  None       Hard  None    Hard  
31596         Hard     hard  None       Hard  None    Hard  

[114 rows x 8 columns]

In [10]:
# The above shows that last_updated column have worng entries
# let us drop these rows
df.dropna(subset=['year'], inplace = True)

In [11]:
df.shape

(34460, 8)

In [12]:
df.isnull().sum()

title           0
author_id       5
last_updated    0
category        0
year            0
date_month      0
month           0
date            0
dtype: int64

In [13]:
df[df["author_id"].isna()]

title author_id  last_updated  \
8234                                        Puzzles       NaN  23 Jul, 2018   
14665                     JavaTuples | Introduction       NaN  16 Sep, 2021   
15073                HTML5 | Mathematical operators       NaN  03 Dec, 2018   
27434  GATE | GATE-CS-2015 (Mock Test) | Question 2       NaN  08 Nov, 2021   
33812                  UwU text convertor in Python       NaN  07 Aug, 2021   

      category   year date_month month date  
8234      easy   2018     23 Jul   Jul   23  
14665    basic   2021     16 Sep   Sep   16  
15073    basic   2018     03 Dec   Dec   03  
27434   medium   2021     08 Nov   Nov   08  
33812   expert   2021     07 Aug   Aug   07

In [14]:
df.dropna(subset=['author_id'], inplace = True)

In [15]:
df.shape

(34455, 8)

In [16]:
df.isnull().sum()

title           0
author_id       0
last_updated    0
category        0
year            0
date_month      0
month           0
date            0
dtype: int64

In [17]:
# Now our data frame is clean 
# We do not have any null values
# Let us now check the dtype of all the columns

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34455 entries, 0 to 34573
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   title         34455 non-null  object
 1   author_id     34455 non-null  object
 2   last_updated  34455 non-null  object
 3   category      34455 non-null  object
 4   year          34455 non-null  object
 5   date_month    34455 non-null  object
 6   month         34455 non-null  object
 7   date          34455 non-null  object
dtypes: object(8)
memory usage: 2.4+ MB


In [19]:
#dtype of the year and date should be integer type
df["year"] = df["year"].astype(int)
df["date"] = df["date"].astype(int)

In [20]:
# Now our dtype is also correct
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34455 entries, 0 to 34573
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   title         34455 non-null  object
 1   author_id     34455 non-null  object
 2   last_updated  34455 non-null  object
 3   category      34455 non-null  object
 4   year          34455 non-null  int64 
 5   date_month    34455 non-null  object
 6   month         34455 non-null  object
 7   date          34455 non-null  int64 
dtypes: int64(2), object(6)
memory usage: 2.4+ MB


In [21]:
df.author_id.value_counts()
# the articles written by GeeksforGeeks are way too large than other authors articles

GeeksforGeeks        11932
ManasChhabra2          317
Striver                261
manjeet_04             246
Chinmoy Lenka          191
                     ...  
KritikaSharma21          1
RatulSarkar1             1
shivamyadav17            1
ksinghshubh              1
sandeep tadepalli        1
Name: author_id, Length: 5583, dtype: int64

In [22]:
# Most frequent author
author_name = df.author_id.value_counts().index[1:11]
author_count = df.author_id.value_counts().values[1:11]
fig = px.histogram(x = author_name, y = author_count)

fig.update_layout(
        xaxis_title = "Author Name",
        yaxis_title = "Article Count",
        title = "<b>Total article per author</b>", # making the title bold
        plot_bgcolor = "#ECECEC")  # changing plot background color

fig.show()

Conclusions:
1. Above graph shows the top 11 authors excluding first.

In [23]:
# check the status of top 10 author other GeeksforGeeks
df_author_other = []

for i in author_name:
    df_new = df[df.author_id == i].groupby(["author_id", "category",
                "year"]).size().reset_index(name = "count").sort_values(by = "count",
                                                            ascending = False)
    df_author_other.append(df_new)
    
frame_other = pd.concat(df_author_other)
    
frame_other.head()

author_id category  year  count
0   ManasChhabra2    basic  2019    135
2   ManasChhabra2    basic  2021    131
1   ManasChhabra2    basic  2020     14
3   ManasChhabra2    basic  2022     13
15  ManasChhabra2   medium  2021      7

In [24]:
# grouped bars

fig = px.bar(frame_other, x = "author_id", y = "count", color = "category",
             barmode='group')

# automatic color according to the category 


fig.update_layout(
        title = "<b>Distribution of articles published by GeeksforGeeks over year</b>",
        plot_bgcolor = "#ECECEC")


fig.show()

Conclusions:
1. ManasChhabra2 published maximum article after GeeksforGeeks
2. Basic level article are popular among the above top 11 (excluding first) authors

In [25]:
# checking the category
df_category = df.groupby("category").size().reset_index(name = "count")
df_category

category  count
0    basic   8127
1     easy   9602
2   expert   2094
3     hard   4223
4   medium  10409

In [26]:
fig = px.bar(df_category, x = "category", y = "count",  text_auto='.2s',
             color_discrete_sequence = ["#ffd514"] * len(df_category))
# providing custom color to the bars in chart
# text indicating the value of the bar

fig.update_layout(
        title = "<b> Distribution of Category </b>",
        plot_bgcolor = "#ECECEC",
        xaxis={'categoryorder':'total descending'} # arranging the bars in descending order
        )
fig.show()

# Note that we did not provide the x-axis and y-axis label.
# The column name is the name for the x axis and y axis label

Conclusions:
1. Maximum articles are of medium category
2. Expert level articles nearly 5 times less that medium level article

In [27]:
df_month = df.groupby(["year","month"]).size().reset_index(name = "count")
df_month.head()

year month  count
0  2010   Nov      1
1  2011   Apr      1
2  2012   Oct      5
3  2013   Feb     48
4  2013   Jan     20

In [28]:
# stack bars

fig = px.bar(df_month, x = "month", y = "count", color = "year")
fig.update_layout(
        title = "<b>Count of Yearly Articles Published per Month</b>",
        plot_bgcolor = "#ECECEC",
        xaxis={'categoryorder':'total descending'})
fig.show()


Conclusions:
1. Maximum articles are published in June
2. The number of articles published in years 2010-2016 are very less
3. The frequency of articles publishing increases from the 2018
4. Articles from 2 months of 2022 (January and Febuary) are also present. This means data is recent

In [29]:
df_author_geek = df[df.author_id == "GeeksforGeeks"].groupby(["author_id", "category",
                "year"]).size().reset_index(name = "count").sort_values(by = "count",
                                                            ascending = False)
df_author_geek.head()

author_id category  year  count
22  GeeksforGeeks     easy  2021   1903
52  GeeksforGeeks   medium  2021   1755
10  GeeksforGeeks    basic  2021   1222
41  GeeksforGeeks     hard  2021    616
50  GeeksforGeeks   medium  2019    594

In [30]:
# grouped bars

fig = px.bar(df_author_geek, x = "year", y = "count", color = "category", barmode='group',
            color_discrete_map= {'basic': '#3C8DD6', 'expert': '#EC2781',
                                "easy":"#008237", "medium":"#FDDA0D", "hard":"#510020"},)
# custom color in bars 

fig.update_layout(
        title = "<b>Distribution of articles published by GeeksforGeeks over year</b>",
        plot_bgcolor = "#ECECEC")


fig.show()

Coclusions:
1. GeeksforGeeks published many articles in 2021
2. Medium and Easy are the most common article category
3. Articles at Expert level are very less

In [31]:
df.head()

title         author_id  \
0        5 Best Practices For Writing SQL Joins       priyankab14   
1                  Foundation CSS Dropdown Menu  ishankhandelwals   
2  Top 20 Excel Shortcuts That You Need To Know       priyankab14   
3                     Servlet – Fetching Result   nishatiwari1719   
4                              Suffix Sum Array          rohit768   

   last_updated category  year date_month month  date  
0  21 Feb, 2022     easy  2022     21 Feb   Feb    21  
1  20 Feb, 2022     easy  2022     20 Feb   Feb    20  
2  17 Feb, 2022     easy  2022     17 Feb   Feb    17  
3  17 Feb, 2022     easy  2022     17 Feb   Feb    17  
4  21 Feb, 2022     easy  2022     21 Feb   Feb    21

In [32]:
# distribution of article as per the bi-weekly in 2021
df_year = df.loc[df["year"] == 2021]

bi_week = []
for i in df_year.date:
    if i <= 15:
        bi_week.append(0)
    else:
        bi_week.append(1)
df_year["bi_weekly"] = bi_week
df_year.head()

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



title          author_id  \
46  Unthinkable Solutions Interview Experience | O...      GeeksforGeeks   
48  Which built-in method returns the calling stri...    mrtwinklesharma   
50  Amazon Internship Interview Experience | On-Ca...  rashimnarayantiku   
51  Goldman Sachs Internship Interview Experience ...      GeeksforGeeks   
52        Spring Boot – CRUD Operations using MongoDB    AmiyaRanjanRout   

    last_updated category  year date_month month  date  bi_weekly  
46  30 Dec, 2021     easy  2021     30 Dec   Dec    30          1  
48  31 Dec, 2021     easy  2021     31 Dec   Dec    31          1  
50  28 Dec, 2021     easy  2021     28 Dec   Dec    28          1  
51  28 Dec, 2021     easy  2021     28 Dec   Dec    28          1  
52  28 Dec, 2021     easy  2021     28 Dec   Dec    28          1

In [33]:
df_2021 = df_year.groupby(["month", "bi_weekly"]).size().reset_index(name = "count")
df_2021.head()

month  bi_weekly  count
0   Apr          0    652
1   Apr          1    903
2   Aug          0    922
3   Aug          1    964
4   Dec          0    690

In [34]:
# dividing data into two columns depending on the week they fall
fig = px.bar(df_2021, x = "month", y = "count", facet_col = "bi_weekly",
            color_discrete_sequence = ["#000080"] * len(df_2021))

# adding a mean line to check the distribution
fig.add_hline(y=df_2021["count"].mean(), line_width=2, line_dash="dash", 
              line_color="#FF9933")

fig.update_layout(
    paper_bgcolor = "#ECECEC",
    plot_bgcolor = "#ECECEC",
    autosize = True,
    yaxis = dict(
        title_text = "<b> Count of Articles </b>",
        titlefont = dict(size = 12) # adding font size for y-axis
    ),
    title_text = "<b> Distribution of Articles Bi-Weekly </b>",
    title_font_size = 16, # adding font size to the article
    title_font_color = "black", # adding font color to the article
    title_pad_t = 5, # adding space to the title heading
    title_pad_l = 20 # adding space to the title heading
    )
 
# adding line color, line width, and tick lables to both the axis
fig.update_yaxes(showticklabels = True, showline = True, linewidth = 2, linecolor = "black")
fig.update_xaxes(showticklabels = True, showline = True, linewidth = 2, linecolor = "black")


Conclusions:
1. The orange line shows the mean of the count of articles published in 2021
2. Most of the month shows that the articles are equally distributed
3. For the month of June we see that the third and fourth week shows too many articles published on the second half